spi: lpspi: rtio: Re-introduce native RTIO support

ubieda opened the pull request:
This version is based on the CPU LPSPI driver, directly implementing
the submit call following the non-blocking asynchronous pattern.

This has been re-introduced after struggling to keep up with data
streaming applications, which demands reduced and controlled latency,
which for the default implementation (using RTIO workqueue) is not
guaranteed, due to being serviced by a thread-pool vs directly in the
ISR context.

This version includes limited feature-set, yet stress-tested:
- Required both rx-fifo and tx-fifo to be equal.
- 8-byte words supported only.
- SPI Controller-only supported (target mode not included).
- SPI_HOLD_ON_CS only supported with CS GPIOs.

ubieda committed:
spi: lpspi: rtio: Re-introduce native RTIO support

This version is based on the CPU LPSPI driver, directly implementing
the submit call following the non-blocking asynchronous pattern.

This has been re-introduced after struggling to keep up with data
streaming applications, which demands reduced and controlled latency,
which for the default implementation (using RTIO workqueue) is not
guaranteed, due to being serviced by a thread-pool vs directly in the
ISR context.

This version includes limited feature-set, yet stress-tested:
- Required both rx-fifo and tx-fifo to be equal.
- 8-byte words supported only.
- SPI Controller-only supported (target mode not included).
- SPI_HOLD_ON_CS only supported with CS GPIOs.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented:
This is the driver to be used by CogniPilot flight controllers. Still a draft as I'm still in the process of stress-testing.

@decsny looping you in since you've been working with LPSPI for a while. Any feedback you may have is welcome.

decsny commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_rtio.c:
```diff
@@ -0,0 +1,515 @@
+
+static void lpspi_iodev_complete(const struct device *dev, int status);
+
+static inline uint8_t rx_fifo_cur_len(LPSPI_Type *base)
+{
+	return (base->FSR & LPSPI_FSR_RXCOUNT_MASK) >> LPSPI_FSR_RXCOUNT_SHIFT;
+}
+
+static inline uint8_t tx_fifo_cur_len(LPSPI_Type *base)
+{
+	return (base->FSR & LPSPI_FSR_TXCOUNT_MASK) >> LPSPI_FSR_TXCOUNT_SHIFT;
+}
```
should be moved to a common file imo since it is the same between this and the other file

ubieda replied:
Moved to common private header file

ubieda committed:
spi: lpspi: Extract FIFO length query helpers to common header file

To remove code duplication.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
spi: lpspi: Add include guards to common header file

Not motivated by any impacted functionality but rather following good
practices. Recursive inclusions may be confusing and hard to diagnose.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

decsny commented on drivers/spi/spi_nxp_lpspi/Kconfig:
```diff
@@ -36,6 +36,17 @@ config SPI_NXP_LPSPI_CPU
+config SPI_NXP_LPSPI_RTIO_CPU
+	bool "NXP LPSPI RTIO-based driver"
+	default y
+	depends on SPI_NXP_LPSPI_CPU
```
im confused about this dependency, does it actually have anything to do with the non-rtio driver? I haven't dug into the driver yet, but I don't understand why it would have this dependency. If it does, it shouldn't, these drivers are intended to be separate and anything shared goes in the common files

decsny replied:
@ubieda as far as I understand what you actually want to do it's to replace the normal CPU driver with the RTIO one when SPI_RTIO is set, which is fine. So just disable the other one when that happens and make this one behave the same way as that one does in regards to dma or not. It makes no sense to make this driver "depend on" the one you are trying not to use.

decsny replied:
> There’s likely benefits in setting a transfer size past which to use DMA or not, atleast I found this to be the case with Atmel SAM where past a certain size the DMA latency cost was offset by not doing lots of software interrupts and the transfer would be done sooner. Maybe also true and a desired attribute for this driver.

Yeah that is a good idea probably but outside the scope for this PR. For now I suggest the author to follow the currently intended pattern. Also, as a side note, this CPU driver also is not optimized for the hardware, due to it's trying to meet the very flexible Zephyr API requirements. I was thinking about making another CPU-based driver that is extremely simpler/smaller and optimized for the hardware. But would have the constraint that it would have to copy the scattered spi_buf into an intermediate buffer and not be able to handle more than either 4K or 65K words (I forget which the hardware supports for max frame size off the top of my head) in a transfer, which probably is fine for most use cases.

decsny replied:
https://github.com/zephyrproject-rtos/zephyr/issues/96116 
https://github.com/zephyrproject-rtos/zephyr/issues/96117

I made these two issues for tracking what we just talked about with more explanation

decsny commented on drivers/spi/spi_nxp_lpspi/CMakeLists.txt:
```diff
@@ -1,5 +1,10 @@
 
 zephyr_library_sources_ifdef(CONFIG_SPI_NXP_LPSPI spi_nxp_lpspi_common.c)
-zephyr_library_sources_ifdef(CONFIG_SPI_NXP_LPSPI_CPU spi_nxp_lpspi.c)
+if(CONFIG_SPI_NXP_LPSPI_RTIO_CPU)
+	zephyr_library_sources(spi_nxp_lpspi_rtio.c)
+else()
+	zephyr_library_sources_ifdef(CONFIG_SPI_NXP_LPSPI_CPU spi_nxp_lpspi.c)
+endif()
```
there should be no if() else() or endif() here

igalloway replied:
Just for my own learning.. should it be the following because they are in Kconfig symbols?

zephyr_library_sources_ifdef(CONFIG_SPI_NXP_LPSPI_RTIO_CPU spi_nxp_lpspi_rtio.c)
zephyr_library_sources_ifdef(CONFIG_SPI_NXP_LPSPI_CPU spi_nxp_lpspi.c)

decsny commented:
@EmilioCBen can you help reviewing the driver since you worked on the previous version

ubieda commented:
I've addressed the comments on Kconfig/CMake selection.
Please take another look @decsny @EmilioCBen, thanks!

dleach02 commented:
@hakehuang can you do a regression test on hardware with this PR and provide feedback?

hakehuang commented:
> @hakehuang can you do a regression test on hardware with this PR and provide feedback?

I kick a rtio spi testing, will feedback once done.

hakehuang reviewed (dismissed):
board test pass on mimxrt1170_evk_cm7 v4.2.0-4311-g3bb21eda08c7.

EmilioCBen commented:
I might be missing something, but is there an established relationship between using rtio and/or a spi device where a transaction is one to one with a clock cycle or a cpu cycle?

decsny reviewed (commented):
some sort of nits but  I didnt review the buffer handling in super details - but I skimmed enough to try to understand generally how it works - correct me if my understanding is wrong because I don't know RTIO much, but I guess each SQE  is corresponding to one spi buffer in the case where `transceive` api is used? and I guess in the case where `transceive` is not called, it's up to some upper layer to define how the bufs are still? so we still have same issue with continuous mode which you are solving same way as other lpspi driver, if I understand right?

Except, if that's correct, then I have  multiple questions:
1) how do we tell when should the CS deassert? right now it seems like the CS deassert when there is no more SQE to process, but how do we know they were all part of the same transfer , that CS should not deassert between some frames. 

2) And also I guess RTIO upper layer handles case of two consumer trying to request to use same spi controller, but where does it prevent same consumer from trying to use the spi controller with two different chip select if the queue of all sqe didnt complete before putting an event with a different CS? because what i imagine is in this case there would be two CS asserted erroneously? but I dont know if this is already handled, so can you please explain

EmilioCBen commented:
If the consonance that the naming scheme is fine than the implementation looks good to me.

ubieda commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_rtio.c:
```diff
@@ -0,0 +1,506 @@
+	for (size_t i = 0; i < fill_len; i++) {
+		base->TDR = 0;
+	}
+}
```
Namespaced all static functions with `lpspi_rtio`

ubieda commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_rtio.c:
```diff
@@ -0,0 +1,506 @@
+
+	lpspi_data->word_size_bytes =
+		DIV_ROUND_UP(SPI_WORD_SIZE_GET(spi_cfg->operation), BITS_PER_BYTE);
+	if (lpspi_data->word_size_bytes != 1) {
+		LOG_ERR("Driver only works with word size = 1 byte");
+		ret = -EINVAL;
+		goto lpspi_iodev_start_on_error;
+	}
+
+	if (op_mode != SPI_OP_MODE_MASTER) {
+		LOG_ERR("Target mode not supported for LPSPI RTIO");
+		ret = -ENOTSUP;
+		goto lpspi_iodev_start_on_error;
+	}
```
Changed to LOG_WRN

ubieda commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_rtio.c:
```diff
@@ -0,0 +1,506 @@
+	}
+	base->TCR &= ~(LPSPI_TCR_CONT_MASK | LPSPI_TCR_CONTC_MASK);
+	/* don't need to wait for TCR since we are at end of xfer + in IRQ context */
+}
```
Got rid of this helper

ubieda commented:
> it probably would have been better if we can somehow make it so that if transceive variants are called by application, then it goes through normal interrupt based path in the other file, but if doing iodev submit, then go through rtio path, so that app isnt forced into one or the other for all spi instances and cases

The one problem I see is that both calls may be performed on the same underlying bus, right? So with this approach we'd need a way to arbitrate calls from both transceive/submit, which is already done for you if you channel all requests through RTIO.
